text/1022-freetypes.rst

```diff
@@ -0,0 +1,213 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2023-01-23
+    Authors: Yury Selivanov <yury@edgedb.com>
+
+
+======================================================
+RFC 1022: Typing free objects & simplifying SDL syntax
+======================================================
+
+Abstract
+========
+
+This RFC proposes enhancements to the EdgeDB type system allowing users to
+explicitly type free objects. Free object types can then be used to type
+function parameters or used to indicate types of query arguments.
+
+While discussing the necessary new syntax for typing free objects, this
+RFC proposes a set of adjustments to the current EdgeDB Schema Declaration
+Language syntax.
+
+
+Typing free objects
+===================
+
+The proposed grammar is structured with the following requirements in mind:
+
+* The syntax must be "light" as it will appear frequently in EdgeQL
+  queries inside cast (e.g. ``<type_expr>$foo``) constructs.
+
+* We want to automatically infer as many things as possible.
+
+* We want the syntax to play nice with the existing syntax for typing
+  arrays and tuples.
+
+The grammar looks as follows::
+
+    type_name:
+        # name of the type, such as "int64" or "default::User"
+
+    type_expr:
+        # type expressions valid in SDL, such as "A | B" or
+        # "array<int64>"
+
+    property_or_link:
+        ["multi" | "single"] ["required" | "optional"] ["property" | "link"] \
+            name ":" type_name | type_expr | free_object_type
+
+    free_object_type:
+        "{"
+            property_or_link
+            [ "," property_or_link ] *
+        "}"
+
+
+Key takeaways:
+
+* EdgeDB will infer whether a field is a property or a link, making
+  "property" and "link" keywords optional.
+
+* Similar to SDL, all properties and links are *optional* and
+  *single* by default.
+
+(Note that the syntax deviates from SDL: we use ``:`` instead of ``->`` and
+``property`` / ``link`` modifier keywords are optional. To make the new free
+object types syntax compatible with SDL we also propose to update the SDL
+syntax in the below `Updates to SDL`_ section.)
+
+Examples:
+
+A free object as a query argument::
+
+    with
+        data := <{a: int64, b: str}>$arg
+    select
+        data.a + <str>data.b;
+
+A more complex free object with nested data and JSON casts::
+
+    with
+        data := <{
+            required pokemon: str,
+            multi required locations: {
+                required point: tuple<float64, float64>,
+                street: str
+            }
+        }>json_array_unpack(<json>$arg)
+
+    for item in data union (
+        insert Pokemon {
+            name := item.pokemon,
+            # ...
+        }
+    )
+
+In user-defined functions::
+
+    function compute(pokemon: {
+        required pokemon: str,
+        multi required locations: {
+            required point: tuple<float64, float64>,
+            street: str
+        }
+    }) -> int64 using (
```

Comment:
You're showing arrow here instead of `:` but declare:
> Replace ``->`` symbol with ``:`` for type annotations of pointers and functions return types.

That said, I would keep an arrow for function return types, because that improves readability. Or do we expect function signatures to be used in type casts anytime soon?

Reply:
> Or do we expect function signatures to be used in type casts anytime soon?

Probably not. :)

> That said, I would keep an arrow for function return types, because that improves readability.

I'm not 100% against that. Rust is using arrows for return types, right? Python uses them too, but I think the main motivation for Python was the same one we have with whitespace-sensitive SDL -- Python can't use `:` for that.

I slightly prefer using `:` instead of `->` for function returns to minimize the surface of ASCII operators we have. If we go the `:` route all in we can just phase out `->` completely and pretty much forget that it exists in all our docs / tutorials / etc.

Reply:
I don't think we can ever get rid of `->` parsed as an operator for backwards compatibility reasons.

From Rust docs:
> ## [2.2 -> for function return type](https://doc.rust-lang.org/1.6.0/complement-design-faq.html#--for-function-return-type)
>
>This is to make the language easier to parse for humans, especially in the face of higher-order functions. fn foo<T>(f: fn(i32): i32, fn(T): U): U is not particularly easy to read.

Well, what is the chance that we will need first-class functions eventually?

Reply:
> Well, what is the chance that we will need first-class functions eventually?

Almost 100% and soon (in 4.0 most likely)

I also think that `->` is a superior syntax for function types and that we should keep it.